src/app-layer-expectation.c

```diff
@@ -0,0 +1,285 @@
+SC_ATOMIC_DECLARE(uint32_t, expectation_count);
+
+#define EXPECTATION_TIMEOUT 30
+
+typedef struct Expectation_ {
+    struct timeval ts;
+    Port sp;
+    Port dp;
+    AppProto alproto;
+    int direction;
+    void *data;
+    struct Expectation_ *next;
+} Expectation;
+
+static void ExpectationListFree(void *e)
+{
+    Expectation *exp = (Expectation *)e;
+    Expectation *lexp;
+    while (exp) {
+        lexp = exp->next;
+        if (exp->data)
+            SCFree(exp->data);
+        SCFree(exp);
+        exp = lexp;
+    }
+}
+
+static void ExpectationDataFree(void *e)
+{
+    if (e)
+        SCFree(e);
+}
+
+uint64_t ExpectationGetCounter(void)
+{
+    uint64_t x = SC_ATOMIC_GET(expectation_count);
+    return x;
+}
+
+void AppLayerExpectationSetup(void)
+{
+    g_expectation_id = IPPairStorageRegister("expectation", sizeof(void *), NULL, ExpectationListFree);
+    g_expectation_data_id = FlowStorageRegister("expectation", sizeof(void *), NULL, ExpectationDataFree);
+    SC_ATOMIC_INIT(expectation_count);
+}
+
+static inline int GetFlowAddresses(Flow *f, Address *ip_src, Address *ip_dst)
+{
+    memset(ip_src, 0, sizeof(*ip_src));
+    memset(ip_dst, 0, sizeof(*ip_dst));
+    if (FLOW_IS_IPV4(f)) {
+        FLOW_COPY_IPV4_ADDR_TO_PACKET(&f->src, ip_src);
+        FLOW_COPY_IPV4_ADDR_TO_PACKET(&f->dst, ip_dst);
+    } else if (FLOW_IS_IPV6(f)) {
+        FLOW_COPY_IPV6_ADDR_TO_PACKET(&f->src, ip_src);
+        FLOW_COPY_IPV6_ADDR_TO_PACKET(&f->dst, ip_dst);
+    } else {
+        return -1;
+    }
+    return 0;
+}
+
+/**
+ * Create an entry in expectation list
+ *
+ * Create a expectation from an existing Flow. Currently, only Flow between
+ * the two original IP addresses are supported.
+ *
+ * \param f a pointer to the original Flow
+ * \param direction the direction of the data in the expectation flow
+ * \param src source port of the expected flow, use 0 for any
+ * \param dst destination port of the expected flow, use 0 for any
+ * \param alproto the protocol that need to be set on the expected flow
+ * \param data pointer to data that will be attached to the expected flow
+ *
+ * \return -1 if error
+ * \return 0 if success
+ */
+
+int AppLayerExpectationCreate(Flow *f, int direction, Port src, Port dst, AppProto alproto, void *data)
+{
+    Expectation *iexp = NULL;
+    IPPair *ipp;
+    Address ip_src, ip_dst;
+
+    Expectation *exp = SCCalloc(1, sizeof(*exp));
+    if (exp == NULL)
+        return -1;
+
+    exp->sp = src;
+    exp->dp = dst;
+    exp->alproto = alproto;
+    exp->ts = f->lastts;
+    exp->data = data;
+    exp->direction = direction;
+
+    if (GetFlowAddresses(f, &ip_src, &ip_dst) == -1)
+        goto error;
+    ipp = IPPairGetIPPairFromHash(&ip_src, &ip_dst);
+    if (ipp == NULL)
+        goto error;
+
+    iexp = IPPairGetStorageById(ipp, g_expectation_id);
+    exp->next = iexp;
+    IPPairSetStorageById(ipp, g_expectation_id, exp);
+
+    SC_ATOMIC_ADD(expectation_count, 1);
+    IPPairUnlock(ipp);
```

Comment:
IPPairRelease?  IPPairGetIPPairFromHash increments use_cnt, IPPairUnlock doesn't decrement.

Reply:
I want to have ref cnt to 1 so it will not be suppressed. The expectation search will decref twice.

Reply:
I think it's best to keep the lookup logic and storing of expectation logic separate for this. With a lookup use_cnt is incremented, and it should be decremented if you are done searching. The storing of an expectation should bump the use_cnt and when it's removed it should be decremented. To avoid the possibility of int overflows it's probably best to increment only when the first is set and decrement only when the last is removed.

Reply:
I'm doing that. Code is not decrementing the counter in the only case of the creation function.